File: frontend/src/types/holding.ts
import { z } from "zod";
import { AssetType, assetTypeSchema } from "./transaction";

// ==================== 資料模型 ====================

/**
 * 持倉明細
 * 對應後端 Holding 結構
 */
export interface Holding {
  symbol: string;
  name: string;
  asset_type: AssetType;
  quantity: number;
  avg_cost: number;
  total_cost: number;
  current_price: number;
  currency: string;
  current_price_twd: number; // TWD 轉換後的價格
  market_value: number;
  unrealized_pl: number;
  unrealized_pl_pct: number;
  price_source?: string; // 價格來源 (cache, api, stale-cache)
  is_price_stale?: boolean; // 價格是否過期
  price_stale_reason?: string; // 價格過期原因
}

/**
 * 持倉篩選條件
 */
export interface HoldingFilters {
  asset_type?: AssetType;
  symbol?: string;
}

// ==================== Zod Schema（用於驗證）====================

/**
 * 持倉 Schema
 */
export const holdingSchema = z.object({
  symbol: z.string(),
  name: z.string(),
  asset_type: assetTypeSchema,
  quantity: z.number(),
  avg_cost: z.number(),
  total_cost: z.number(),
  current_price: z.number(),
  currency: z.string(),
  current_price_twd: z.number(),
  market_value: z.number(),
  unrealized_pl: z.number(),
  unrealized_pl_pct: z.number(),
  price_source: z.string().optional(),
  is_price_stale: z.boolean().optional(),
  price_stale_reason: z.string().optional(),
});

/**
 * 持倉篩選條件 Schema
 */
export const holdingFiltersSchema = z.object({
  asset_type: assetTypeSchema.optional(),
  symbol: z.string().optional(),
});

// ==================== 輔助函式 ====================

/**
 * 格式化貨幣顯示
 * @param value 金額
 * @param currency 貨幣代碼
 * @returns 格式化後的字串
 */
export function formatCurrency(value: number, currency: string): string {
  const prefix = currency === "TWD" ? "NT$" : currency === "USD" ? "$" : "";
  return `${prefix} ${value.toLocaleString("zh-TW", {
    minimumFractionDigits: 0,
    maximumFractionDigits: 2,
  })}`;
}

/**
 * 格式化百分比顯示
 * @param value 百分比數值
 * @returns 格式化後的字串
 */
export function formatPercentage(value: number): string {
  const sign = value >= 0 ? "+" : "";
  return `${sign}${value.toFixed(2)}%`;
}

/**
 * 取得損益顏色類別（台灣股市習慣：紅漲綠跌）
 * @param value 損益數值
 * @returns Tailwind CSS 類別名稱
 */
export function getProfitLossColor(value: number): string {
  return value >= 0 ? "text-red-600" : "text-green-600";
}

/**
 * 計算總市值
 * @param holdings 持倉陣列
 * @returns 總市值
 */
export function calculateTotalMarketValue(holdings: Holding[]): number {
  return holdings.reduce((sum, h) => sum + h.market_value, 0);
}

/**
 * 計算總成本
 * @param holdings 持倉陣列
 * @returns 總成本
 */
export function calculateTotalCost(holdings: Holding[]): number {
  return holdings.reduce((sum, h) => sum + h.total_cost, 0);
}

/**
 * 計算總損益
 * @param holdings 持倉陣列
 * @returns 總損益
 */
export function calculateTotalProfitLoss(holdings: Holding[]): number {
  return holdings.reduce((sum, h) => sum + h.unrealized_pl, 0);
}

/**
 * 計算總損益百分比
 * @param holdings 持倉陣列
 * @returns 總損益百分比
 */
export function calculateTotalProfitLossPct(holdings: Holding[]): number {
  const totalCost = calculateTotalCost(holdings);
  const totalProfitLoss = calculateTotalProfitLoss(holdings);
  return totalCost > 0 ? (totalProfitLoss / totalCost) * 100 : 0;
}

/**
 * 按資產類型分組持倉
 * @param holdings 持倉陣列
 * @returns 分組後的持倉
 */
export function groupHoldingsByAssetType(
  holdings: Holding[]
): Record<AssetType, Holding[]> {
  return holdings.reduce((groups, holding) => {
    const type = holding.asset_type;
    if (!groups[type]) {
      groups[type] = [];
    }
    groups[type].push(holding);
    return groups;
  }, {} as Record<AssetType, Holding[]>);
}

/**
 * 排序持倉
 * @param holdings 持倉陣列
 * @param sortBy 排序欄位
 * @param order 排序順序
 * @returns 排序後的持倉陣列
 */
export function sortHoldings(
  holdings: Holding[],
  sortBy: "market_value" | "unrealized_pl" | "quantity" | "symbol",
  order: "asc" | "desc" = "desc"
): Holding[] {
  const sorted = [...holdings].sort((a, b) => {
    let compareValue = 0;
    switch (sortBy) {
      case "market_value":
        compareValue = a.market_value - b.market_value;
        break;
      case "unrealized_pl":
        compareValue = a.unrealized_pl - b.unrealized_pl;
        break;
      case "quantity":
        compareValue = a.quantity - b.quantity;
        break;
      case "symbol":
        compareValue = a.symbol.localeCompare(b.symbol);
        break;
    }
    return order === "asc" ? compareValue : -compareValue;
  });
  return sorted;
}

/**
 * 搜尋持倉
 * @param holdings 持倉陣列
 * @param query 搜尋關鍵字
 * @returns 符合條件的持倉陣列
 */
export function searchHoldings(holdings: Holding[], query: string): Holding[] {
  if (!query) return holdings;
  const lowerQuery = query.toLowerCase();
  return holdings.filter(
    (h) =>
      h.symbol.toLowerCase().includes(lowerQuery) ||
      h.name.toLowerCase().includes(lowerQuery)
  );
}

File: backend/internal/models/holding.go
package models

import (
	"time"
)

// Holding 持倉資料
// 代表某個標的（symbol）的當前持倉狀況
// 所有金額欄位（成本、市值、損益）統一以 TWD 計價
type Holding struct {
	Symbol           string    `json:"symbol"`                       // 標的代碼（例如：2330, AAPL, BTC）
	Name             string    `json:"name"`                         // 標的名稱
	AssetType        AssetType `json:"asset_type"`                   // 資產類型
	Quantity         float64   `json:"quantity"`                     // 當前持有數量
	AvgCost          float64   `json:"avg_cost"`                     // FIFO 計算的平均成本（含手續費，TWD）
	AvgCostOriginal  float64   `json:"avg_cost_original"`            // FIFO 計算的平均成本（含手續費，原幣別）
	TotalCost        float64   `json:"total_cost"`                   // 總成本 = AvgCost * Quantity（TWD）
	CurrentPrice     float64   `json:"current_price"`                // 當前市場價格（原始幣別）
	Currency         Currency  `json:"currency"`                     // 價格幣別
	CurrentPriceTWD  float64   `json:"current_price_twd"`            // 當前市場價格（TWD）
	MarketValue      float64   `json:"market_value"`                 // 市值 = CurrentPriceTWD * Quantity（TWD）
	UnrealizedPL     float64   `json:"unrealized_pl"`                // 未實現損益 = MarketValue - TotalCost（TWD）
	UnrealizedPLPct  float64   `json:"unrealized_pl_pct"`            // 未實現損益百分比
	LastUpdated      time.Time `json:"last_updated"`                 // 最後更新時間
	PriceSource      string    `json:"price_source,omitempty"`       // 價格來源（cache, api, stale-cache）
	IsPriceStale     bool      `json:"is_price_stale,omitempty"`     // 價格是否過期
	PriceStaleReason string    `json:"price_stale_reason,omitempty"` // 價格過期原因
}

// CostBatch FIFO 成本批次
// 用於追蹤每一批買入的成本，實作 FIFO 計算
// 成本統一以 TWD 計價
type CostBatch struct {
	Date             time.Time `json:"date"`               // 買入日期
	Quantity         float64   `json:"quantity"`           // 該批次剩餘數量
	UnitCost         float64   `json:"unit_cost"`          // 單位成本（含手續費，TWD）
	UnitCostOriginal float64   `json:"unit_cost_original"` // 單位成本（含手續費，原幣別）
	OriginalQty      float64   `json:"original_qty"`       // 原始買入數量
	Currency         Currency  `json:"currency"`           // 原始交易幣別
	ExchangeRate     float64   `json:"exchange_rate"`      // 交易時的匯率（TWD/原幣別）
}

// HoldingFilters 持倉篩選條件
type HoldingFilters struct {
	AssetType *AssetType `json:"asset_type,omitempty"` // 按資產類型篩選
	Symbol    *string    `json:"symbol,omitempty"`     // 按標的代碼篩選
}

// CorporateActionType 公司行動類型（股票分割/合併）
type CorporateActionType string

const (
	ActionTypeSplit CorporateActionType = "split" // 股票分割（例如 1:2）
	ActionTypeMerge CorporateActionType = "merge" // 股票合併（例如 2:1）
)

// CorporateAction 公司行動記錄
// 用於處理股票分割、合併等事件
type CorporateAction struct {
	Symbol string              `json:"symbol"` // 標的代碼
	Type   CorporateActionType `json:"type"`   // 行動類型
	Date   time.Time           `json:"date"`   // 生效日期
	Ratio  float64             `json:"ratio"`  // 比例（例如 1:2 分割 = 2.0, 2:1 合併 = 0.5）
}

// Validate 驗證 CorporateActionType 是否有效
func (c CorporateActionType) Validate() bool {
	switch c {
	case ActionTypeSplit, ActionTypeMerge:
		return true
	}
	return false
}

